crypto/sha/asm/sha512-x86_64.pl

```diff
@@ -276,34 +280,54 @@ ()
 .cfi_startproc
 ___
 $code.=<<___ if ($SZ==4 || $avx);
-	lea	OPENSSL_ia32cap_P(%rip),%r11
-	mov	0(%r11),%r9d
-	mov	4(%r11),%r10d
-	mov	8(%r11),%r11d
+    lea OPENSSL_ia32cap_P(%rip),%r10
+    mov 0(%r10),%r9
+    mov 8(%r10),%r11d
 ___
-$code.=<<___ if ($SZ==4 && $shaext);
-	test	\$`1<<29`,%r11d		# check for SHA
-	jnz	_shaext_shortcut
+$code.=<<___ if ($SZ==8);
+    mov 20(%r10),%r10d
 ___
-$code.=<<___ if ($avx && $SZ==8);
-	test	\$`1<<11`,%r10d		# check for XOP
-	jnz	.Lxop_shortcut
+if ($SZ==4) {
+$code.=<<___ if ($shaext);
+    test \$`1<<29`,%r11d            # check for SHA
+    jnz  _shaext_shortcut
 ___
 $code.=<<___ if ($avx>1);
-	and	\$`1<<8|1<<5|1<<3`,%r11d	# check for BMI2+AVX2+BMI1
-	cmp	\$`1<<8|1<<5|1<<3`,%r11d
-	je	.Lavx2_shortcut
+    and \$`1<<8|1<<5|1<<3`,%r11d    # check for BMI2+AVX2+BMI1
+    cmp \$`1<<8|1<<5|1<<3`,%r11d
+    je  .Lavx2_shortcut
+___
+} else { # $SZ==8
+$code.=<<___ if ($avx);
+    bt \$43,%r9                     # check for XOP
+    jc .Lxop_shortcut
+___
+if ($avx>1) { # $SZ==8 && $avx>1
+$code.=<<___;
+    test \$`1<<5`,%r11d             # check for AVX2
+    jz   .Lavx_dispatch
+___
+$code.=<<___ if ($SZ==8);
+    test \$`1<<0`,%r10d             # AVX2 confirmed, check SHA512
```

Comment:
With https://github.com/openssl/openssl/pull/25709 `OPENSSL_ia32cap_P` was extended from 2 `uint64_t`s to 5 and feature flags corresponding to SHA512/SM3/SM4 capabilities are copied to the 3rd `uint64_t` at bit indices:
- 32 for SHA512 ISA extension
- 33 for SM3 ISA extension
- 34 for SM4 ISA extension

At this line `%r10d` points to `(uint8_t *)OPENSSL_ia32cap_P + 20`.

A `uint32_t` at this location will indicate the following capabilities:
- bit 0 (lsb) is SHA512
- bit 1 is SM3
- bit 2 is SM4

@nhorman could you please let me know if I've answered your question.